### ISSUE TITLE
[Prometheus] Prometheus Reporter Implementation

### ISSUE THREAD

**Issue by Rurouni** [alhardy/AppMetrics.Reporters#6]
Hi, are there any plans to support Prometheus?
Easiest option may be to try to adapt [Metrics.NET.Prometheus](https://github.com/DEXSinc/Metrics.NET.Prometheus). 
[prometheus-net](https://github.com/andrasm/prometheus-net) is another option but will probablty require much more effort.

**Comment by alhardy** [comment]
Yes definitely, although I'm wanting to get a v1 release out first with InfluxDB, same as https://github.com/alhardy/AppMetrics/issues/67

promotheus-net looks like an option however it sounds like they may be a waiting for dotnet core 2.0 due to a dependency on HttpListener - https://github.com/andrasm/prometheus-net/issues/31

The other option would be to [write a client](https://prometheus.io/docs/instrumenting/writing_clientlibs/) targeting dotnet standard which like it would be very similar to the [InfluxDB client](https://github.com/alhardy/AppMetrics.Reporters/tree/dev/src/App.Metrics.Extensions.Reporting.InfluxDB)

**Comment by Rurouni** [comment]
I has built prototype implementation [App.Metrics.Extensions.Prometheus](https://github.com/Rurouni/App.Metrics.Extensions.Prometheus)  and whilst counters/meters work well,  the problem lies in Histograms/Timers:
-  First(and the blocker) they don't expose lifetime sum of observed values so there is no way to satisfy minimal requirement for _count/_sum timeseries  
    -  see https://prometheus.io/docs/concepts/metric_types/#summary
    - at the moment I am using count*mean but that hack won't work with decaying reservoir, so if you can expose it as part of HistogramValue that would be great
-  Second (not critical) they could be mapped only to prometheus summary(precalculated quantiles) but not histogram (pure cumulative buckets) and that will cause issues when aggregating over multiple nodes later
    -  see https://prometheus.io/docs/practices/histograms/ and http://latencytipoftheday.blogspot.co.uk/2014/06/latencytipoftheday-you-cant-average.html

There are also questions about how Tags/Labels should be correctly applied. At the moment your library assumes tags (both key and value) are applied once on metrics creation + there is special case for Meters with additional Mark(count, subItem). With such approach it is quite problematic to create single metric for example HttpRequest { route:"/api/whatever", method:"GET", statusCode:200} where additional dimensions are encoded with Tags/Labels.  It is possible at the moment by creating multiple metric instances but then there is no way to correlate them back to single metric for purpose of exporting to Prometheus. I was thinking that whilst you don't have v1 release it may be worth to create better API around multidimensional metrics, like for example:
```
IMeter m = ...Provider.Create(...TagDefinitions...)
m.WithTags("Tag1Val", "Tag2Val").Mark()
```
Thank you

**Comment by alhardy** [comment]
@Rurouni  awesome, that's a great start.

> First(and the blocker) they don't expose lifetime sum of observed values so there is no way to satisfy minimal requirement for _count/_sum timeseries
> - see https://prometheus.io/docs/concepts/metric_types/#summary
> - at the moment I am using count*mean but that hack won't work with decaying reservoir, so if you can expose it as part of HistogramValue that would be great

By **_sum** we're referring to the sum of observed values within current sample set?

> It is possible at the moment by creating multiple metric instances but then there is no way to correlate them back to single metric for purpose of exporting to Prometheus.

Yes, see [here](https://github.com/alhardy/AppMetrics/blob/dev/src/App.Metrics.Extensions.Middleware/Extensions/MetricsExtensions.cs#L50) for example, the `EndpointHttpErrorRequests` Meter ha a `Group` set on the metric options, then in the influxdb reporter, the [group name is reported as the metric name](https://github.com/alhardy/AppMetrics.Reporters/blob/dev/src/App.Metrics.Extensions.Reporting.InfluxDB/Extensions/PackMetricValueSourceExtensions.cs#L67) and the name of the meter reported as a tag. The [docs](https://alhardy.github.io/app-metrics-docs/getting-started/fundamentals/organizing-metrics.html#metric-grouping) on this need more work.

**Comment by Rurouni** [comment]
> By _sum we're referring to the sum of observed values within current sample set?

They say **total sum**  in their documentation, so I assume it should be monotonically incrementing sum of all observed values since metric creation. 

Prometheus can handle rare metric data reset when process restarts but not on every reporting period otherwise many functions like rate() will be broken.
That's why current mapping of AppMetrics types to Prometheus is following:
- gauge -> gauge
- counter -> gauge (AppMetrics counters can go up and down)
- meter -> counter (only count needed because rate can be calculated in Prometheus)
- histogram -> summary (count as _count, _sum broken, percentiles-> 0.x quantiles)
- timer -> summary (same as histogram, because there is no need for additional Meter data as _count is enough to calculate rate later)

**Comment by alhardy** [comment]
OK let me think about it, that sum could get very large very quickly!

Let me know how you go with Metric grouping

**Comment by Rurouni** [comment]
> OK let me think about it, that sum could get very large very quickly!

Interestingly, I have just looked at  prometheus-net implementation and all they do is `_sum += value;` where `_sum` is double. Clearly at some stage it will have FP error more than value and it will stop changing at all but on other hand double can account for 16 digits and that's a lot (for example assuming 100 000 RPS and observing values 0.2s latency for a **year** it will lose precision up to 0.001)

> Let me know how you go with Metric grouping

It is possible to do it this way. Slightly more complicated than with influxDb or elasticsearch because prometheus requires multiple metrics with same name but different labels to be packaged together. So in this case metric name becomes "unique_name,tag1:val1|tag2:val2" and group is actual metric name.
What worries me is that it all feels very ad hoc and allows too many concepts: Context, Group, Mark(MetricItem, val), Mark(string, val), MetricTags and then there is a lot of conventions between metric definitions and specific reporter implementation.  All modern reporting options (InfluxDB, ElasticSearch, Prometheus, OpenTSDB) work really nicely with such metric structure  `<<full metric name>> { [string immutableLabel1, string value1], [string immutableLabel2, string value2], ... }` where labels are rarely changed and values often but from limited set(not completely unbounded).

**Comment by alhardy** [comment]
Re Metric Group / Context / Tags, yeah I feel the same way, however at the moment I can't think of a better way. If tags are set at run-time under the same metric name how would we sample/aggregate data in memory before flushing at an interval? 

Also we wouldn't be able to represent that in the json endpoint. What I'm thinking of having is a UI where one can register endpoints and an agent which can collect metrics for the registered endpoints, deserialise and then flush.

**Comment by Rurouni** [comment]
I have updated https://github.com/Rurouni/App.Metrics.Extensions.Prometheus to group multidimensional metrics based on what you proposed before with `Group`, The only thing that doesn't support it is Gauge, because I couldn't figure how can I just set Gauge to specific value without hacking messy lambda function as valueProvider (It would be great to have option in API to just call `Gauge.SetValue(double)` )

Also as example of what I was suggesting before. That's what I am using atm:
https://gist.github.com/Rurouni/42577acbc6f9af0a28cb30ae8da25cb9

**Comment by alhardy** [comment]
@Rurouni Sorry haven't gotten back to you yet, I'll need until early next week. Initial thoughts are:

What you've done is already supported for counters and meters e.g. 

`_metrics.Measure.Counter.Decrement(MetricOptions.Counter.TestCounter, new MetricSetItem(keys, values));`

The above will also give you a total count of all metrics within the set plus percentages per item etc.

But yes no support for multi-dimensional timers etc at the moment, let me think about it for a bit.

**Comment by alhardy** [comment]
Protobuf serialization is a nice idea for those using a pull based approach, is the usual approach taken with Prometheus?

**Comment by Rurouni** [comment]
> is the usual approach taken with Prometheus?

Prometheus will happily work with both so it really depends on how many metrics do you have. More metrics bigger win from proto.

> What you've done is already supported for counters and meters e.g.

I know, that was my original implementation, but then I have switched to using multiple metrics instead for several reasons:
1. I don't need separate total or percentages because it can be calculated in prometheus/grafana 
1. MetricTags in subitems don't include global tags automatically, so more work for me
1. To have single way of representing grouped metrics across timers, meters, counters, etc

Will you be able to add non `Func<double>` based Gauges and  total non-decaying sum property in timers any time soon? (alternatively if you can upgrade to latest RC msbuild dotnet tooling/VS2017 I think I can do it myself)

**Comment by alhardy** [comment]
OK yeah totally agree re multi-dimensional metrics, I will add it in, but yeah will need until early next week unfortunately, and yes I will look at the other items you've requested. These are next on my list.

VS2017 is out March 07 I think, thats on my list too ;)

If you'de like it sooner yeah fork the project and get it started, that would be great

**Comment by alhardy** [comment]
Multidimensional Metrics: https://github.com/alhardy/AppMetrics/commit/a976979c858116546bacbab72aad7afe491cc048

**Comment by Rurouni** [comment]
Wow, that's one huge commit, nice work 👍 
One thing I hasn't understood when crawling through code is what is the story for reporting side? Are they just different metrics with specific name structure and it's up to reporter to join them into groups, or they will be already somehow grouped when reporter runs?

**Comment by alhardy** [comment]
I've removed Groups altogether, as you pointed out it's confusing 👍 I still feel it needs some work but it's working for now, I may make more changes.

On the reporting side, you can look at the influxdb [reporter changes](https://github.com/alhardy/AppMetrics.Reporters/commit/aa8841cbe9a9ed434b0dcd9aee7115d13763777d)

**Comment by alhardy** [comment]
Your thoughts on having another metric type for `Summary` or just part of `Histogram`? Prometheous identify there is a difference. I still need to read into this more.

https://github.com/alhardy/AppMetrics/issues/91

**Comment by Rurouni** [comment]
Ah, I see, I missed the fact that original name in MetricValueSourceBase remains not joined with tags.
Cool, it should be straightforward then to adapt Prometheus reporter to new approach.  Though I have something working at the moment with your old version, so I think I will wait for your final 1.0 release before switching

**Comment by Rurouni** [comment]
The difference between Summary and Histogram is that summary represents already calculated quantiles (like what App.Metrics does for timers) whilst histogram is just  buckets with counts(like 0-100ms requests count, 100-200ms request count etc.).. the benefit of having latest is that you can have statistics sane reaggregation of percentiles across multiple server instances...the downside is that user needs to predetermine buckets knowing distribution of values in his metric.

Because App.Metrics model is much closer to StatsD then Prometheus I was using same conversion as in https://github.com/prometheus/statsd_exporter

**Comment by alhardy** [comment]
OK thanks, that makes sense re histograms and summaries.

In regards to your other two requests:

- Support `Gauge.SetValue(double)`: what's your use case here? My thought  is it makes more sense being lazy by using a func, invoking only when flushing, e.g.` () => GetCpuUsage()` we would want to evaluate that when reporting otherwise when would we set the value?
- Support lifetime sum of all observes values on a histogram: Timers at the moment have a `TotalTime` property, does that satisfy your use case with prometheous? In regards to tracking the sum on a histogram I'm thinking it may in some circumstances result in an `ArithmeticException` depending on what's being tracked and the measurement unit, I guess in these cases I could set the value to `double.PositiveInfinity`

**Comment by Rurouni** [comment]
- My use case is that I maintain 2 backends for metrics (through App.Metrics and DogstatsD) and gauges in latter are just send double. For AppMetrics first I though to just pass ()=>val every time somebody calls SetValue(double) but because value factory is registered only once on Gauge creation that won't work.
- If `TotalTime` is non-resetting sum of all time observations then it would close my use case completely because I don't use bare histograms anywhere only timers.

**Comment by alhardy** [comment]
`sum` is now exposed on Histograms - https://github.com/alhardy/AppMetrics/commit/e2d96262d2dd65870aca2c569a919068a49c4eb6

`TotalTime` is a non-resetting sum, since sum is exposed in Histograms now I'll remove `TotalTime`

**Comment by alhardy** [comment]
`TotalTime` removed from Histograms and `Gauge.SetValue(double)` is now supported

**Comment by Rurouni** [comment]
Hi @alhardy ,
I have almost fully ported prometheus reporter to recent v1 release. Only issue that I have run into is with timers/histograms.
I am not sure if it's do with timers in general or just my usage of it but, the problem is that in Histogram instance behind Timer is actually getting shared across metrics with different dimensions. For example if I have metrics like `RequestTimer { path: /x } and RequestTimer { path: /y }` then they will have equal values for Histogram component even despite being reported as different metrics to reporter. Interestingly enough Meter component is working as expected, also Histogram.LastValue was different

I am using timers in such form:
```
_metrics.Provider.Timer.Instance(_definition, new MetricTags(_dimensions, tagValues))
                    .Record((long)value.TotalMilliseconds, TimeUnit.Milliseconds);
```
because I need to record already calculated time and I couldn't find way to pass that information through .Measure.Timer interface.

**Comment by alhardy** [comment]
@Rurouni That was quick! 

Hmm, yeah that shouldn't be the case, I'll have to investigate and get back to you.

**Comment by Rurouni** [comment]
@alhardy  I have figured out the reason for that bug.
In my wrapper around App.Metrics I was doing that:
```
_metrics.Provider.Timer.Instance(_definition, new MetricTags(_dimensions, tagValues))
                    .Record((long)value.TotalMilliseconds, TimeUnit.Milliseconds);
```
The problem is that _definition is actually defined once per my metric instance and reused for different dimensions.
```
_definition = new TimerOptions()
            {
                Name = metricName,
                RateUnit = TimeUnit.Seconds,
                DurationUnit = TimeUnit.Seconds,
                Reservoir = new Lazy<IReservoir>(() => new DefaultForwardDecayingReservoir())
            };
```
and it looks like reservoir is shared across multiple dimensions. 
The question is how do you correctly override reservoir for multidimensional metrics?

**Comment by alhardy** [comment]
@Rurouni I don't think I've see that, check the sanbox project in the main repo https://github.com/alhardy/AppMetrics#grafanainfluxdb-web-monitoring

Need to confirm though, could you log this as in issue in the [main repo](https://github.com/alhardy/AppMetrics/issues)

**Comment by alhardy** [comment]
Fixed here https://github.com/alhardy/AppMetrics/issues/107

**Comment by Rurouni** [comment]
It's fixed in upcoming 1.1 release right?

**Comment by alhardy** [comment]
Yep, I've made some [changes around metrics serialization](https://github.com/alhardy/AppMetrics/issues/106) as well, would be great if you could check that out in supporting your [prometheous PR](https://github.com/alhardy/AppMetrics.Reporters/pull/16) before releasing the next version.

**Comment by Rurouni** [comment]
I'll try to find some time this week to update it

**Comment by Rurouni** [comment]
Ok, creating `IMetricsResponseWriter` was pretty straightforward but I've run into another issue.
At the moment there is no way to enable just `MetricsEndpointMiddleware` without enabling also the whole lot of tracking middlewares:
```csharp
app.UseMiddleware<ActiveRequestCounterEndpointMiddleware>();
app.UseMiddleware<ErrorRequestMeterMiddleware>();
app.UseMiddleware<PerRequestTimerMiddleware>();
app.UseMiddleware<OAuthTrackingMiddleware>();
app.UseMiddleware<PostAndPutRequestSizeHistogramMiddleware>();
app.UseMiddleware<RequestTimerMiddleware>();
app.UseMiddleware<ApdexMiddleware>();
```
 In my case I have bespoke tracing/tracking middlewares, so having additional 7 middlewares on every request would be just needless performance hit.

**Comment by alhardy** [comment]
OK, I'll add an option to the startup config allowing you to set `options.RegisterDefaultMetricsMiddleware = false` which will ignore all app metrics middleware related to tracking. I can include that in 1.1.0 stable. Sound OK?

**Comment by alhardy** [comment]
Created [Prometheus repo](https://github.com/alhardy/AppMetrics.Extensions.Prometheus/) with boilerplate and added you as a collaborator @Rurouni

**Comment by alhardy** [comment]
Implementation [here](https://github.com/alhardy/AppMetrics.Extensions.Prometheus/)
